File: lib/features/payments/ui/history_screen_data.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import '../data/payment.dart';

part 'history_screen_data.freezed.dart';

@freezed
abstract class PaymentListItemData with _$PaymentListItemData {
  const factory PaymentListItemData({
    required String id,
    required String billerName,
    required String account,
    required int amountPaise,
    required DateTime paidAt,
    required PaymentStatus status,
  }) = _PaymentListItemData;
}

@freezed
sealed class HistoryScreenData with _$HistoryScreenData {
  const factory HistoryScreenData.empty() = HistoryEmpty;
  const factory HistoryScreenData.loaded({
    required List<PaymentListItemData> items,
  }) = HistoryLoaded;
}

File: src/features/payments/ui/historyScreenData.ts
import { useStore } from 'zustand';
import type { PaymentStatus, Payments } from '../data/payment';
import { useContainer } from '../../../core/container/containerContext';
export interface PaymentListItemData {
  readonly id: string;
  readonly billerName: string;
  readonly account: string;
  readonly amountPaise: number;
  readonly paidAt: Date;
  readonly status: PaymentStatus;
}
export type HistoryScreenData =
  | { readonly kind: 'empty' }
  | { readonly kind: 'loaded'; readonly items: readonly PaymentListItemData[] };

export const projectHistory = (payments: Payments): HistoryScreenData => {
  if (payments.items.length === 0) return { kind: 'empty' };
  return {
    kind: 'loaded',
    items: [...payments.items].reverse().map((p) => ({
      id: p.id,
      billerName: p.billerName,
      account: p.account,
      amountPaise: p.amountPaise,
      paidAt: p.paidAtUtc,
      status: p.status,
    })),
  };
};

export const useHistoryScreenData = (): HistoryScreenData => {
  const container = useContainer();
  const payments = useStore(container.payments, (s) => s.payments);
  return projectHistory(payments);
};
